keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/AdminService.kt

```diff
@@ -32,4 +34,27 @@ class AdminService(
     log.info("Triggering a recheck for $resourceId by clearing our record of the diff")
     diffFingerprintRepository.clear(resourceId)
   }
+
+  /**
+   * Removes the stored state we have for any stateful constraints for an environment
+   * so they will evaluate again
+   */
+  fun reevaluateConstraints(application: String, environment: String, type: String? = null) {
+    log.info("[app=$application, env=$environment] Triggering reevaluation of stateful constraints.")
+    if (type != null) {
+      log.info("[app=$application, env=$environment] Triggering only type $type")
+    }
+
+    val deliveryConfig = repository.getDeliveryConfigForApplication(application)
+    val env = deliveryConfig.environments.find { it.name == environment }
+      ?: throw UserException("No environment $environment in config for $application")
```

Comment:
Maybe add a `NoSuchEnvironment` sub-class and add it to the list of exceptions that cause a 404 to return in `ExceptionHandler`?

Reply:
You're right, I was just being lazy. I'll do that